front/app/containers/Admin/pages-menu/containers/HeroBannerForm/index.tsx

```diff
@@ -0,0 +1,492 @@
+  signedOutHeaderErrors: Multiloc;
+  signedOutSubheaderErrors: Multiloc;
+};
+export type PreviewDevice = 'mobile' | 'tablet' | 'desktop';
+
+// resources
+import {
+  IHomepageSettingsAttributes,
+  updateHomepageSettings,
+} from 'services/homepageSettings';
+import useHomepageSettings from 'hooks/useHomepageSettings';
+
+// utils
+import { isNil, isNilOrError } from 'utils/helperUtils';
+import { isCLErrorJSON } from 'utils/errorUtils';
+import { forOwn, size, trim, debounce, pick } from 'lodash-es';
+import { convertUrlToUploadFile } from 'utils/fileUtils';
+
+// constants
+import { pagesAndMenuBreadcrumb, homeBreadcrumb } from '../../breadcrumbs';
+const TITLE_MAX_CHAR_COUNT = 45;
+const SUBTITLE_MAX_CHAR_COUNT = 90;
+
+const BgHeaderPreviewSelect = styled(Select)`
+  margin-bottom: 20px;
+`;
+
+const HeroBannerForm = ({ intl: { formatMessage } }: InjectedIntlProps) => {
+  const theme: any = useTheme();
+  const homepageSettings = useHomepageSettings();
+
+  // component state
+  const [isLoading, setIsLoading] = useState(false);
+  const [headerLocalDisplayImage, setHeaderLocalDisplayImage] = useState<
+    UploadFile[] | null
+  >(null);
+  const [apiErrors, setApiErrors] = useState<CLError[] | null>(null);
+  const [headerAndSubheaderErrors, setHeaderAndSubheaderErrors] =
+    useState<MultilocErrorType>({
+      signedOutHeaderErrors: {},
+      signedOutSubheaderErrors: {},
+    });
+  const [bannerError, setBannerError] = useState<string | null>(null);
+  const [localHomepageSettings, setLocalHomepageSettings] =
+    useState<IHomepageSettingsAttributes | null>(null);
+  const [previewDevice, setPreviewDevice] = useState<PreviewDevice>('desktop');
+
+  useEffect(() => {
+    if (isNilOrError(homepageSettings)) {
+      return;
+    }
+
+    // copy homepage settings to local state
+    setLocalHomepageSettings({
+      ...homepageSettings.data.attributes,
+      banner_layout:
+        homepageSettings.data.attributes.banner_layout ??
+        'full_width_banner_layout',
+    });
+
+    // the image file sent from the API needs to be converted
+    // to a format that can be displayed. this is done locally
+    // when the image is changed but needs to be done manually
+    // to process the initial API response
+    const convertHeaderToUploadFile = async (fileInfo) => {
+      if (fileInfo) {
+        const tenantHeaderBg = await convertUrlToUploadFile(fileInfo.url);
+        const headerBgUploadFile = !isNilOrError(tenantHeaderBg)
+          ? [tenantHeaderBg]
+          : [];
+        setHeaderLocalDisplayImage(headerBgUploadFile);
+        setBannerError(null);
+      }
+    };
+
+    const headerFileInfo = homepageSettings.data.attributes.header_bg?.large;
+    convertHeaderToUploadFile(headerFileInfo);
+  }, [homepageSettings]);
+
+  const onSave = async () => {
+    if (localHomepageSettings) {
+      // only send the keys that relate to the banner
+      // and layout back to the API
+      const bannerRelevantKeys = [
```

Comment:
We use the whole `homePageSettings` object here, but I didn't want to send the whole thing back to the API to overwrite the current values with a `PATCH`, so I'm picking out only the keys used on this page and sending those back. Not sure if this is the best solution, would like it to match the TS types but I'm not sure how. Here are some possibilities:

* Ignore this and send the entire `homePageSettings` object back to the server
* Filter out the relevant settings for this page when they're set from the API
* Somehow link this to TS types to reduce possibility of bugs/typos

Reply:
More generally, your question is how we deal with diffs, right? This is a question you could ask in #chapter-front-end. 

We sometimes have a diff object in the state that contains the values to update. Search for `IUpdatedProjectProperties` to see examples.

Reply:
LMK if you need help.

Reply:
We had a diff system in the last version of this component and it was pretty complicated, I don't think full diffing is necessary here. Basically the `homepageSettings` object has 20 properties, but we will only ever update 6 of them in this form. I don't want to send all 20 back to the API when the user saves this form, 

I think I could do a simple diff between the `homepageSettings` coming from the API and whatever's been updated  in state, which in this form would only ever be those 6 properties, and then we wouldn't need to list them explicitly